pkg/cloudclient/aws/private.go

```diff
@@ -312,6 +312,145 @@ func getMasterNodeSubnets(kclient client.Client) (map[string]string, error) {
+
+	var publicSubnets []string
+
+	machineList, err := baseutils.GetMasterMachines(kclient)
+
+	if err != nil {
+		log.Error(err, "No master machines found")
+		return nil, err
+	}
+
+	// Get the first master machine in the list
+	masterMachine := machineList.Items[0]
+
+	// Get the AWS ProviderSpec from the machine
+	machineProviderSpec, err := getAWSDecodedProviderSpec(masterMachine)
+	if err != nil {
+		log.Error(err, "Cannot decode AWS Provider Spec")
+		return nil, err
+	}
+
+	// Get the subnet the master machine is in
+	initalSubnet := machineProviderSpec.Subnet.ID
+
+	// Look up the subnet metadata
+	describeSubnetOutput, err := c.ec2Client.DescribeSubnets(&ec2.DescribeSubnetsInput{SubnetIds: []*string{initalSubnet}})
+
+	// Extract the VPC ID from the subnet metadata
+	targetVPC := describeSubnetOutput.Subnets[0].VpcId
+
+	// List all subnets in the VPC
+	allSubnets, err := c.getAllSubnetsInVPC(*targetVPC)
+
+	// List all route tables associated with the VPC
+	routeTables, err := c.getAllRouteTablesInVPC(*targetVPC)
+
+	for _, subnet := range allSubnets {
+		isPublic, err := isSubnetPublic(routeTables, *subnet.SubnetId)
+
+		if err != nil {
+			log.Error(err, "Error while determining if subnet is public")
+			return nil, err
+		}
+		if isPublic {
+			publicSubnets = append(publicSubnets, *subnet.SubnetId)
+		}
+	}
+
+	return publicSubnets, nil
+}
+
+func (c *Client) getAllSubnetsInVPC(vpcID string) ([]*ec2.Subnet, error) {
+
+	var subnetIDs []*ec2.Subnet
+	token := aws.String("initString")
+
+	for token != nil {
+		describeSubnetOutput, err := c.ec2Client.DescribeSubnets(&ec2.DescribeSubnetsInput{Filters: []*ec2.Filter{{Name: aws.String("vpc-id"), Values: []*string{aws.String(vpcID)}}}})
+		if err != nil {
+			log.Error(err, "Error while descirbing subnets")
+			return nil, err
+		}
+		for _, subnet := range describeSubnetOutput.Subnets {
+			subnetIDs = append(subnetIDs, subnet)
+		}
+
+		token = describeSubnetOutput.NextToken
+	}
+
+	return subnetIDs, nil
+}
+
+func (c *Client) getAllRouteTablesInVPC(vpcID string) ([]*ec2.RouteTable, error) {
+
+	var routeTables []*ec2.RouteTable
+	token := aws.String("initString")
+
+	for token != nil {
+		describeRouteTablesOutput, err := c.ec2Client.DescribeRouteTables(&ec2.DescribeRouteTablesInput{Filters: []*ec2.Filter{{Name: aws.String("vpc-id"), Values: []*string{aws.String(vpcID)}}}})
+		if err != nil {
+			log.Error(err, "Error while describing route tables")
+			return nil, err
+		}
+		for _, routeTable := range describeRouteTablesOutput.RouteTables {
+			routeTables = append(routeTables, routeTable)
+		}
+
+		token = describeRouteTablesOutput.NextToken
+	}
+
+	return routeTables, nil
+}
+
+func isSubnetPublic(rt []*ec2.RouteTable, subnetID string) (bool, error) {
+	var subnetTable *ec2.RouteTable
+	for _, table := range rt {
+		for _, assoc := range table.Associations {
+			if aws.StringValue(assoc.SubnetId) == subnetID {
+				subnetTable = table
+				break
+			}
+		}
+	}
```

Comment:
What's going on here? Should the `break` on line 414 be breaking out of the outermost loop?

Reply:
Yes and no. This function specifically was copied directly from the installer, as mentioned in the PR description. While breaking out of the outer loop is probably better here, as it stands it still works, as a subnet can only be associated with one table. I wanted to avoid changing the upstream function as much as possible, so I left it as is.

Reply:
Okay, that makes sense. I wish upstream would break better, but that's not for this PR. ;-)

Other changes look good.